### ISSUE TITLE
Text overlaps on other other text or even images

### ISSUE THREAD

**Issue by ghost** [openzim/mwoffliner#1004]
Most issues can be seen in titles. Affects all range of Wikis.

http://library.kiwix.org/wikipedia_gsw_all_nopic_2019-06/A/Wikipedia:Houptsyte
GSW
http://library.kiwix.org/wikibooks_gl_all_nopic_2019-07/A/Portada
GL
http://library.kiwix.org/wikipedia_scn_all_nopic_2019-04/A/P%C3%A0ggina_principali
SCN
http://library.kiwix.org/wikinews_bs_all_nopic_2019-07/A/Po%C4%8Detna_strana
BS
http://library.kiwix.org/wikipedia_tt_all_nopic_2019-04/A/%D0%91%D0%B0%D1%88_%D0%B1%D0%B8%D1%82
TT
http://library.kiwix.org/wikinews_bs_all_novid_2019-07/A/Po%C4%8Detna_strana
BS
http://library.kiwix.org/wikipedia_eu_all_novid_2019-06/A/Azala
EU
http://library.kiwix.org/wikisource_pt_all_novid_2019-07/A/Wikisource:P%C3%A1gina_principal
PT
http://library.kiwix.org/wikiversity_fr_all_nopic_2019-07/A/Wikiversit%C3%A9:Accueil
FR
http://library.kiwix.org/granbluefantasy_en_all_all_novid_2018-10/A/Main_Page.html
EN
http://library.kiwix.org/wikipedia_tt_all_novid_2019-04/A/%D0%91%D0%B0%D1%88_%D0%B1%D0%B8%D1%82
TT
http://library.kiwix.org/wikipedia_vec_all_nopic_2019-04/A/Pajina_prinsipa%C5%82e
VEC
http://library.kiwix.org/wikipedia_it_all_nopic_2019-10/A/Pagina_principale
IT
http://library.kiwix.org/wikipedia_lez_all_nopic_2019-05/A/%D0%9A%D1%8C%D0%B8%D0%BB%D0%B8%D0%BD_%D1%87%D1%87%D0%B8%D0%BD
LEZ
http://library.kiwix.org/wikipedia_xal_all_nopic_2019-04/A/%D0%9D%D2%AF%D1%80_%D1%85%D0%B0%D0%BB%D1%85
XAL
http://library.kiwix.org/wikipedia_ce_all_novid_2019-04/A/%D0%9A%D0%BE%D1%8C%D1%80%D1%82%D0%B0_%D0%B0%D0%B3%D3%80%D0%BE
CE
http://library.kiwix.org/wikipedia_sr_all_nopic_2018-10/A/%D0%93%D0%BB%D0%B0%D0%B2%D0%BD%D0%B0_%D1%81%D1%82%D1%80%D0%B0%D0%BD%D0%B0.html
SR
http://library.kiwix.org/wikipedia_gan_all_nopic_2019-06/A/%E5%B0%81%E9%9D%A2
GAN
http://library.kiwix.org/wikisource_gl_all_novid_2019-07/A/Portada
GL
http://library.kiwix.org/wikipedia_fy_all_nopic_2019-06/A/Haadside
FY
http://library.kiwix.org/wikipedia_os_all_nopic_2019-04/A/%D0%A1%C3%A6%D0%B9%D1%80%D0%B0%D0%B3_%D1%84%D0%B0%D1%80%D1%81
OS
http://library.kiwix.org/wikipedia_stq_all_nopic_2019-03/A/Haudsiede
STQ
http://library.kiwix.org/wikipedia_pms_all_novid_2019-04/A/Intrada
PMS
http://library.kiwix.org/wikisource_pt_all_nopic_2019-07/A/Wikisource:P%C3%A1gina_principal
PT
http://library.kiwix.org/wikinews_ca_all_nopic_2019-07/A/Portada
CA
http://library.kiwix.org/wikipedia_pms_all_nopic_2019-04/A/Intrada
PMS
http://library.kiwix.org/wikipedia_ce_all_nopic_2019-04/A/%D0%9A%D0%BE%D1%8C%D1%80%D1%82%D0%B0_%D0%B0%D0%B3%D3%80%D0%BE
CE
http://library.kiwix.org/wikisource_fi_all_novid_2019-07/A/Etusivu
FI
http://library.kiwix.org/wikiversity_it_all_nopic_2019-07/A/Pagina_principale
IT
http://library.kiwix.org/wikipedia_frr_all_nopic_2019-06/A/Wikipedia:Hoodsid
FRR
http://library.kiwix.org/wikipedia_lo_all_novid_2019-05/A/%E0%BB%9C%E0%BB%89%E0%BA%B2%E0%BA%AB%E0%BA%BC%E0%BA%B1%E0%BA%81
LO

**Comment by kelson42** [comment]
@benoit74 We still have the problem, but not sure this is a bug.

Lets take https://browse.library.kiwix.org/content/wikiversity_it_all_maxi/Pagina_principale, it looks good... but https://browse.library.kiwix.org/content/wikiversity_it_all_nopic/Pagina_principale has indeed the problem

**Comment by benoit74** [comment]
I think this is a bug. The problem is that since we completely remove the image in nopic, if the layout assumes that the image will be there and consume some "space", then the UI gets broken. Would be interesting to investigate if instead of completely removing the image we can replace it with a block which consumes the same space. 

I think it is quite important because here we have a case where the UI is completely broken, but nopic ZIM is always a bit clunky in its layout due to these missing images. Having something consumes space and even visually indicate that an image was supposed to be here but was removed (with a textured background for instance) would be a clear added value IMHO.

Something like this:

<img width="747" height="316" alt="Image" src="https://github.com/user-attachments/assets/de425cd3-6210-443b-b01b-8048d2428f17" />

```html
<div style="height:160px;width: 213px;background: repeating-linear-gradient( 45deg, #ccc, #ccc 5px, #ddd 5px, #ddd 10px );" class="mw-file-element" id="mwBTc"></div>
```

**Comment by kelson42** [comment]
@benoit74 What about keeping the CSS images in nopic? Would this solve the problem?

**Comment by benoit74** [comment]
Nope, it would not solve the problem of https://browse.library.kiwix.org/content/wikiversity_it_all_nopic/Pagina_principale, these are "real" images coming from the wikitext, not from the CSS.

**Comment by kelson42** [comment]
@benoit74 What would be a proper approach to solve this issue? Do we have a strategy in mind?

**Comment by Markus-Rost** [comment]
At the minimum we should respect the `alt` attributes of images when deleting them, inserting the `alt` value as text. This alone would solve the issue of https://browse.library.kiwix.org/content/wikiversity_it_all_nopic/Pagina_principale and they are only using `alt="&nbsp;"`. Though this is unlikely to solve all formatting issues caused by missing images.

Ideally we would insert some placeholder element taking up the same space as the image, like mentioned above:
```html
<img alt="" src="//upload.wikimedia.org/wikipedia/commons/8/8c/Bluebg_rounded_croped.png" decoding="async" width="250" height="25" class="mw-file-element" data-file-width="250" data-file-height="25">
```
```html
<span class="mw-file-element" style="display: inline-block; width: 250px; height: 25px;"></span>
```

**Comment by benoit74** [comment]
@Markus-Rost you are right regarding `alt`, this should be respected when it does exist.

Moving `alt` text to a `span` (or anything else) is probably going to be clunky because we might have special CSS rules attached to the `<img>` tag.

My strategy would be:
- if `alt` is missing or empty, replace `src` value with "something" (can it be a transparent 1x1 PNG?) and add a CSS class to the image which will create a "patterned" background like on https://github.com/openzim/mwoffliner/issues/1004#issuecomment-3041070103 so that it is clear that something was supposed to be here (we do not end-up with just a blank space and no clue)
- if `alt` has content, replace `src` value with a constant `missing-B3jfbIJEliFQXYxvBXKhGw.png`, where `B3jfbIJEliFQXYxvBXKhGw` is just a random string to be sure this file will never exist, and keep the rest of the `<img>` as-is ; this will allow the browser to detect that image is missing and display the alt text just like it would in "reality"

**Comment by Markus-Rost** [comment]
Unsure about the patterned background. If the image is positioned above or below other elements, this might make some text unreadable.

Unsure about linking the missing image as well. It means the browser will always display a missing file icon before the alt text.

**Comment by KunalSiyag** [comment]
I am not sure about my approach but what if we generate a blank svg that has the alt text in it?  
That fulfil the conditions of showing the alt text as well as has the same img tag (so no CSS layout errors). We just change the src to the new svg generated.

**Comment by benoit74** [comment]
An SVG with the alt text in it has the same flaw than the patterned background, it might hide things or make things unreadable if the image is positioned above or below other elements

**Comment by KunalSiyag** [comment]
If the original image did not break layout, the SVG replacement will not either.  I tried to recreate it

<img width="400" height="400" alt="Image" src="https://github.com/user-attachments/assets/2c1a7d96-9a53-488e-ae1f-c191606d2081" />

I am not sure about how to fill the gap though. If no alt text is present, how to show the distinction on what was supposed to be there. 

One idea is:

- Always preserve the original width and height.
- Show a subtle border only for sufficiently large images.
- Render centered alt text only when it exists and is meaningful.
- For small icons or decorative images, avoid visible borders to prevent visual clutter.

Currently what looks a bit okayish is what if we do something like make the border visible and alt text appears if the image has something like that. 


<img width="500" height="500" alt="Image" src="https://github.com/user-attachments/assets/a2e61482-fcd8-4e9b-beb4-903bcef6c221" />

Maybe now we can make conditions like only show the border if the image is large in size or something like that to reduce those lines entangling with other boxes. It preserves the original layout of the site.

**Comment by benoit74** [comment]
- what if the image has some transparency and something was below it, shown through the transparent zone?
- what if the image has some blank area and some text was above it in the blank area?

**Comment by KunalSiyag** [comment]
If the image had some transparency and it was replaced by the svg, the text would still be there but we will have to see if the image also contained some alt. That would create a problem. 

Even for the image having a blank area, it would be something like the text will stay at its place and the layout will be maintained. The background of the text may be gone depending on what color the image's blank area was.

Similary I tried to create different cases. The problem only arises when we try to show alt. 

To address this, I’m considering two approaches:

The '**On-Demand**' View: Implement a small, clickable arrow button that reveals the alt text only when the user explicitly requests it. 

The **Streamlined Flow**: Omit the alt text entirely in the UI, relying on the visual integrity of the SVG to convey the information.


<img width="757" height="939" alt="Different cases of " src="https://github.com/user-attachments/assets/0410a6ef-4a6c-4365-a265-19ffc9506a81" />

**Comment by Markus-Rost** [comment]
I like the idea of the transparent placeholder images with the alt text. This solves the issue neatly with minimal disruption for the reader.

> * what if the image has some transparency and something was below it, shown through the transparent zone?
> * what if the image has some blank area and some text was above it in the blank area?

These seem like rare cases, that shouldn't be any issue for the huge majority of images. And the slight conflict due to possible overlapping alt text is to me less of an issue compared to the broken layout which exists currently.

**Comment by KunalSiyag** [comment]
I can start working on generating a placeholder for missing images where the original `<img>` is preserved but the src is replaced with a generated SVG placeholder. If alt text is available, it can be displayed within the placeholder.

One **potential edge case** is very long alt text. In such cases we may need to wrap or truncate the text so that it fits reasonably within the image dimensions.

Does this approach sound acceptable? If so, I can start implementing it.

**Comment by benoit74** [comment]
@KunalSiyag thank you, this seems indeed acceptable, I didn't totally understood the SVG approach you proposed, thank you for the "demo" and additional propositions.

I feel like the proposition of your last comment is indeed acceptable, at least it is worth to implement and run on a real wiki and solve any bug /edge case we oversight (if any).

Please propose a PR.
